### ISSUE TITLE
How to apply different decorators on same type

### ISSUE THREAD

**Issue by ChristianHoe** [simpleinjector/SimpleInjector#167]
I'm using commands and queries, based on interfaces, and have batch registered those. All pretty much taken from Steven's [command](https://www.cuttingedge.it/blogs/steven/pivot/entry.php?id=91) / [query](https://www.cuttingedge.it/blogs/steven/pivot/entry.php?id=92) blogs and the examples in the SimpleInjector documentation:

``` c#
interface IQuery<TResult> {}
interface IQueryHandler<TQuery, TResult> where TQuery : IQuery<TResult> {}
```

This is registered with the container like this:

``` c#
container.Register(typeof(IQueryHandler<,>), new[] { typeof(IQueryHandler<,>).Assembly });
container.RegisterDecorator(IQueryHandler<,>), typeof(ValidationDecorator<,>));
```

Now I am looking for a nicer way to apply different decorators depending on the requested Interface like:

``` c#
container.Register(new[] { 
    typeof(IInternalQuerier<,>), 
    typeof(IExternalWpfQuerier<,>),
    typeof(IExternalWcfQuerier<,>) }, 
    new[] { typeof(IBusinessQuerier<,>).Assembly }, 
    Lifestyle.Scoped);
container.RegisterDecorator(IExternalWcfQuerier<,>), typeof(LoggingDecorator<,>));
container.RegisterDecorator(IExternalWpfQuerier<,>), typeof(ExceptionDecorator<,>));
container.RegisterDecorator(IInternalQuerier<,>), typeof(ValidationDecorator<,>));
```

With the definitions:

``` c#
public interface IInternalQuerier<TQuery, TResult> : IQuerier<TQuery, TResult> 
    where TQuery : IQueryRequest<TResult>
public interface IExternalWpfQuerier<TQuery, TResult> : IQuerier<TQuery, TResult> 
    where TQuery : IQueryRequest<TResult>
public interface IExternalWcfQuerier<TQuery, TResult> : IQuerier<TQuery, TResult> 
    where TQuery : IQueryRequest<TResult>

public interface IBusinessQuerier<TQuery, TResult> : 
    IInternalQuerier<TQuery, TResult>,  IExternalWpfQuerier<TQuery, TResult>,
    IExternalWcfQuerier<TQuery, TResult>
    where TQuery : IQueryRequest<TResult>
```

My workaround at the moment is using Adapters like:

``` c#
public class InternalAdapter<TQuery, TResult> : 
    IInternalQuerier<TQuery, TResult> 
    where TQuery : IQueryRequest<TResult>
{ ... public InternalAdapter(IBusinessQuerier<TQuery, TResult> querier) ...}
            
container.Register(typeof(IInternalQuerier<,>), typeof(InternalAdapter<,>), Lifestyle.Scoped);
```

Which just add another Layer, which feels strange/wrong to me.

So basically I am looking for a simple(r) way to offer the same query with a different set of decorators to other components. What is the (best) way to go?

**Comment by qujck** [comment]
why do you have

```csharp
public interface IInternalQuerier<TQuery, TResult> : IQuerier<TQuery, TResult> where TQuery : IQueryRequest<TResult>
public interface IExternalWpfQuerier<TQuery, TResult> : IQuerier<TQuery, TResult> where TQuery : IQueryRequest<TResult>
public interface IExternalWcfQuerier<TQuery, TResult> : IQuerier<TQuery, TResult> where TQuery : IQueryRequest<TResult>
```

what do they give you over and above?

```csharp
interface IQueryHandler<TQuery, TResult> where TQuery : IQuery<TResult> {}
```

**Comment by dotnetjunkie** [comment]
I have the same question as Peter has: why? Might it be you are over complicating things?

**Comment by ChristianHoe** [comment]
> Might it be you are over complicating things?
probably

I using IQueryHandler<> for my Business layer behind a WCF service. For simplicity I have two decorators Validation and Logging.
So my registration was as simple as 
`container.Register(typeof(IQueryHandler<,>), new[] { typeof(IQueryHandler<,>).Assembly });`
`container.RegisterDecorator(IQueryHandler<,>), typeof(ValidationDecorator<,>));`
`container.RegisterDecorator(IQueryHandler<,>), typeof(LoggingDecorator<,>));`

At the beginning I had lots queries getting simple chuncks of data, which I offered via the WCF Service. Now I wanted to extend my WCF service by some methods offering composite data, to reduce the number of calls. Therefor I created new QueryHandler, which uses some of the existing QueryHandler. But for these calls I did not wanted the LoggingDecorator applied.

So I created the interface IInternalQuerier<,> and IExternalWcfQuerier<,> which are combined in IBusinessQuerier<,> to have something to differ at the decorator registration:
`container.Register(typeof(IBusinessQuerier<,>), new[] { typeof(IBusinessQuerier<,>).Assembly }, Lifestyle.Scoped);`
`...`
`container.RegisterDecorator(IExternalWcfQuerier<,>), typeof(LoggingDecorator<,>));`
`container.RegisterDecorator(IExternalWcfQuerier<,>), typeof(ValidationDecorator<,>));`
`...`
`container.RegisterDecorator(IInternalQuerier<,>), typeof(ValidationDecorator<,>));`

Because I got a Torn LifeCycle message when I tried this:
`container.Register(typeof(IExternalWcfQuerier<,>), new[] { typeof(IQueryHandler<,>).Assembly }, Lifestyle.Scoped);`
`container.RegisterDecorator(IExternalWcfQuerier<,>), typeof(LoggingDecorator<,>));`
`container.RegisterDecorator(IExternalWcfQuerier<,>), typeof(ValidationDecorator<,>));`
`...`
`container.Register(typeof(IInternalQuerier<,>), new[] { typeof(IQueryHandler<,>).Assembly }, Lifestyle.Scoped);`
`container.RegisterDecorator(IInternalQuerier<,>), typeof(ValidationDecorator<,>));`

**Comment by qujck** [comment]
I wondered if that might be the case: it's a common problem, one I ran into myself enough times to come up with the term [Holistic Abstractions](http://qujck.com/commands-and-queries-are-holistic-abstractions/).

I would suggest that you do not allow one `QueryHandler` to reference another, if you need to reference more than one then you need a [Mediator](https://en.wikipedia.org/wiki/Mediator_pattern). A `IMediatingHandler` enables you to wrap one or more `QueryHandlers` in a consistent *Holistic Abstraction*.

```csharp
IMediatingHandler<TQuery, TResult> where TQuery : IQuery<TResult> { }
```

```csharp
public class CompositeQuery1 : IQuery<CompositeResult1> { }

public class CompositeMediator1<CompositeQuery1, CompositeResult1> :        
    IMediatingHandler<CompositeQuery1, CompositeResult1> 
{
    private readonly IQueryHandler<Query1, Result1> handler1;
    private readonly IQueryHandler<Query2, Result2> handler2;

    //...

}
```

```csharp
public class OpenGenericSingleQueryMediator<TQuery, TResult> :
    IMediatingHandler<TQuery, TResult> where TQuery : IQuery<TResult>
{
    private readonly IQueryHandler<TQuery, TResult> handler1;

    //...

}
```

Does this make sense?

**Comment by TheBigRic** [comment]
So if I understood correctly you only introduced a new set of abstraction to differ in the applied decorators?

In that case: loose this abstractions, you don't them. You missed a feature which is included in the Simple Injector API. It's descibed [here](http://simpleinjector.readthedocs.org/en/latest/advanced.html#using-contextual-information-inside-decorators) in the documentation.

There is an overload available in which you can access the context in which the decorator is applied. You can use this context to let Simple Injector know if the decorator has to be applied or not.

You could for example decorate the queryhandlers which you want to be logged with a custom attribute `RequiresLogging` and register your decorator as follows:

``` c#
container.RegisterDecorator(typeof(IQueryHandler<,>), typeof(LoggingDecorator<,>),
   context => Attribute.IsDefined(context.ImplementationType, typeof(RequiresLoggingAttribute)));
```
Furthermore the decorator subsysteem is also aware of generic type contraints you place on the open (or partially closed) generic decorator implementation. This means the decorator will only be applied if the generic type constraints are met.

**Comment by ChristianHoe** [comment]
@qujck 
Thanks for link. I need a littlebit time to check, if it really solves my problem.

@TheBigRic 
If I understood you correct, this allows you to apply decorators depending on "properties" of the decorated type. This is not what I want. I like it to be dependend on where it is injected. 
Simplified example: 
* Usage inside of the assembly -> Basis-Decorator-Set applied 
* Usage ouside of the assembly -> Full-Decorator-Set applied

**Comment by dotnetjunkie** [comment]
> This is not what I want. I like it to be dependend on where it is injected. 

You can do this as well:

    container.RegisterDecorator(typeof(IQueryHandler<,>), 
        typeof(LoggingDecorator<,>),
        c => c.ImplementationType.Assembly == typeof(WcfLayerType).Assembly);

I agree both with Peter Ric; use a mediator and ditch the extra abstractions. Apply decorators conditional. This makes your life easier and makes your code cleaner and more maintainable.

**Comment by TheBigRic** [comment]
>  I like it to be dependend on where it is injected. 
> * Usage inside of the assembly -> Basis-Decorator-Set applied 
> * Usage ouside of the assembly -> Full-Decorator-Set applied

I don't understand what you mean with this. You're hiding your businesslayer behind WCF, like [this](https://www.cuttingedge.it/blogs/steven/pivot/entry.php?id=95)? So what is inside the assembly and what is outside...?
With this setup, client side you will typically have one open generic QueryHandler which just forwards the query to WCF.

If I understand the question correctly and you use the WCF setup from the linked blog post, the `IQueryHandlers<,>` implementations on the serverside would be root types and won't get injected in anything! 

Maybe you are injecting one QueryHandler into another? There is no (straightforward) way of knowing if a QueryHandler is being injected or being retrieved from the container as a root type. And frankly it wouldn't make sense if these 2 `GetInstance()` calls would result in different object graphs.

So can you elaborate a bit more?

**Comment by ChristianHoe** [comment]
My simplified project structure is like:
WFC
`GetQuery1(...) { return Business.GetQuery1(...) }`
`GetQuery2(...) { return Business.GetQuery2(...) }`
`GetComposite1(...) { return Business.GetComposite1(...) }`
Business
`GetQuery1(...) : IQueryHandler<,> { ... }`
`GetQuery2(...) : IQueryHandler<,> { ... }`
`GetComposite1(...) : IQueryHandler<,> { ... Business.GetQuery1(...) ... Business.GetQuery2(...) ... }`

I am basically reusing the `GetQuery1(...)` and `GetQuery2(...)` in `GetComposite1(...)`.

The registration is like:
`container.Register(typeof(IQueryHandler<,>), new[] { typeof(IQueryHandler<,>).Assembly }, lifestyle);`
`container.RegisterDecorator(typeof(IQueryHandler<,>), typeof(ValidationDecorator<,>), lifestyle);`
`container.RegisterDecorator(typeof(IQueryHandler<,>), typeof(SecurityDecorator<,>), lifestyle);`
 
So far so easy. But how do I register a `WCFLoggingDecorator<>`? 
Using `container.RegisterDecorator(typeof(IQueryHandler<,>), typeof(LoggingDecorator<,>), lifestyle);` causes multiple calls when calling `GetComposite1(...)` which I do not want. 

That was the moment  were my mess started (see first post).

I could use @qujck mediator in that way, that I inject `IMediatingHandler<>` in the WCF Project and apply the WCF specific decorators on it. But what if I get for example another Web project which needs a different decorator set?

**Comment by TheBigRic** [comment]
So you're not using the maintenance free approach in your WCF project? This is a completely different subject, but you really should reconsider if you will proceed this way. Your basic setup behind WCF is good with using command- and queryhandlers, but exposing them piece by piece on your WCF layer is a maintenance hell. All that is assuming that the connecting client is also maintained by you. If you have to take third party developers into account, the way you're working now is probably the best, but will still make your WCF service a massive project.

That being said, back to your problem. The problem is your current design, exactly as @qujck describes in his blog about holistic abstractions. So the solution is, use the `Mediator`! And you also would work with this same `IMediatingHandler` in your future webproject. I fail to see why that would make any difference. As @qujck already mentioned in this post 
 > All Controller actions (or whatever MVC, MVVM, MVP part for that matter) should reference Mediators not Queries.

What he means is: Everywhere you fire a query you should **not** inject `IQueryHandler` but `IMediatingHandler`. `IQueryHandler` only will be injected in the `IMediatingHandler` implementations. If you're using a `IQueryProcessor` as a kind of dispatcher this would also create `IMediatingHandler`

For the queries you want to call directly (GetQuery1) you will only need a single open generic implementation of `IMediatingHandler` as @qujck mentioned:
``` c#
public class OpenGenericSingleQueryMediator<TQuery, TResult> :
    IMediatingHandler<TQuery, TResult> where TQuery : IQuery<TResult>
{
    private readonly IQueryHandler<TQuery, TResult> handler1;

    public TResult Handle(TQuery query)
    {
        return this.handler1.Handle(query);
    }
}
```

Now you only decorate the `IMediatingHandler` with the `ValidationDecorator`, `SecurityDecorator` and `LoggingDecorator` and you'll be fine.

**Comment by ChristianHoe** [comment]
First of all, thank you for taking your time.

I can not use the maintenance free approach, because my project is a kind of Data Hub for some other systems out of my range. Webservices is a must, so I though `IQueryHandler` is a nice way to go.

The solution with the `IMediatingHandler` looks like just adding another abstraction layer to me. But maybe I just need a small hint.
> And you also would work with this same IMediatingHandler in your future webproject. I fail to see why that would make any difference.

Having change my configuration to `IMediatingHandler` (instead of `IQueryHandler`) for the WCF project like:
`container.Register(typeof(IMediatingHandler<,>), new[] { typeof(IMediatingHandler<,>).Assembly }, lifestyle);`
`container.RegisterDecorator(typeof(IMediatingHandler<,>), typeof(ValidationDecorator<,>), lifestyle);`
`container.RegisterDecorator(typeof(IMediatingHandler<,>), typeof(SecurityDecorator<,>), lifestyle);`

How do I register my `IMediatingHandler` for the web project with a `FancySecurityDecorator` instead of a my normal `SecurityDecorator` ?

**Comment by TheBigRic** [comment]
> How do I register my  IMediatingHandler  for the web project with a  FancySecurityDecorator  instead of a my normal  SecurityDecorator  ?

So what you're basically saying is:
If I connect from WPF client to the WCF service to get some data I need a different level of security than when I connect from my Web (MVC of something) to my WCF service. 

That does not make sense at all. Your WCF service will expose himself to the outside world no matter what client, so security should always be a big issue.

**Comment by ChristianHoe** [comment]
> Well that's exactly what the blog about holistic abstractions is telling you! You miss an abstraction!

Seeing a technical solution and really really understanding not only the technical but also the abstract concept behind it, takes some time (at least for me). I did not came here because I really wanted to do something SOLID with CQRS, but found that blog post about decorators and simple injector, just showing how to solve some of my problems in a yery interessting way. 
Sorry, for not being so fast on the conceptual part.

> If you mean that you will reuse the business layer in your web project than there is no problem at all, because the web project needs it's own composition root no matter what, so you would just register this FancySecurityDecorator . That's the whole point of dependency injection and SOLID design, composing the application from small building blocks.

Good point. Problem solved.

One last problem. Next to my Business layer I have a Dataaccess layer. Because the "application" contains mostly read operation I though about skipping the normal `IRepository` approach and using `IQueryHandler<,>` for most of the time. Being able to use a LoggingDecorator here looks like a plus too. But how to register the decorator here, if I want a Dataaccess specific logging or what ever? I do not see a new composition root as an option and `IQueryHandler` (via Business project) and `IMediatingHandler` (via WCF project) are already decorated/used.

**Comment by qujck** [comment]
My suggestion would be to use a variation of IQueryHandler<,> for data operations, something along the lines of IDataQueryHandler<,>.

**Comment by ChristianHoe** [comment]
Now I have the following structure:

Common
`interface IQuery<TResult>` 
`interface IQueryHandler<TQuery, TResult>`

DataAccess
`interface IDataQueryHandler<TQuery, TResult> : IQueryHandler<TQuery, TResult>`

`class GetQuery1 : IDataQueryHandler<TQuery, TResult>`
`class GetQuery2 : IDataQueryHandler<TQuery, TResult>`

Business
`internal interface IBusinessQueryHandler<TQuery, TResult> : IQueryHandler<TQuery, TResult>`
`interface IMediatingHandler<TQuery, TResult> : IQueryHandler<TQuery, TResult>`
`class OpenGenericSingleQueryMediator<TQuery, TResult> : IMediatingHandler<TQuery, TResult>`

`class GetQuery1 : IBusinessQueryHandler<TQuery, TResult>`
`class GetQuery2 : IBusinessQueryHandler<TQuery, TResult>`
`class Composite1 : IMediatingHandler<TQuery, TResult>`

And the registration looks like:
DataAccess
`container.Register(typeof(IDataQueryHandler<,>), new[] { typeof(IDataQueryHandler<,>).Assembly }, lifestyle);`
`container.RegisterDecorator(typeof(IDataQueryHandler<,>), typeof(ValidationDecorator<,>), lifestyle);`

Business
`container.Register(typeof(IBusinessQuerier<,>), new[] { typeof(IBusinessQuerier<,>).Assembly }, lifestyle);`
`container.RegisterDecorator(typeof(IBusinessQuerier<,>), typeof(ValidationDecorator<,>), lifestyle);`

`container.Register(typeof(IMediatingHandler<,>), new[] { Assembly.GetExecutingAssembly() }, lifestyle);`
`container.RegisterConditional(typeof(IMediatingHandler<,>), typeof(OpenGenericSingleQueryMediator<,>), lifestyle, c => !c.Handled);`

WCF
`container.RegisterDecorator(typeof(IMediatingHandler<,>), typeof(ValidationDecorator2<,>), lifestyle);`
`container.RegisterDecorator(typeof(IMediatingHandler<,>), typeof(LoggingDecorator2<,>), lifestyle);`
`container.RegisterDecorator(typeof(IMediatingHandler<,>), typeof(ExceptionHandlingDecorator2<,>), lifestyle);`

Now it is looking good and clear for me.
- I can decorate DataAccess and Business separately
- I have a clear interface with own decoration possibilities when reusing my business logic (through `IMediatingHandler` and container root)
- Simple Injector is doing the boring stuff automatically

Is there also a nice way to supply standard implementation decorators? At the moment I use inheritance like 
`LoggingDecorator<TQuery, TResult> : Common.LoggingDecorator<TQuery, TResult>, IBusinessQuerier<TQuery, TResult> where TQuery : IQueryRequest<TResult>
{`
`public LoggingQuerierDecorator(IBusinessQuerier<TQuery, TResult> decorated) : base(decorated) {}`
`}`
which just looks wrong.

**Comment by qujck** [comment]
I suggest you keep the interfaces separate:

```csharp
public interface IQuery<TResult> { }
public interface IQueryHandler<TQuery, TResult> where TQuery : IQuery<TResult>
{
    TResult Handle(TQuery query);
}
public interface IDataQueryHandler<TQuery, TResult> where TQuery : IQuery<TResult>
{
    TResult Handle(TQuery query);
}
public interface IMediatingHandler<TQuery, TResult> where TQuery : IQuery<TResult>
{
    TResult Handle(TQuery query);
}
```

And keep the decorators separate:

```csharp
public class QueryHandlerLogger<TQuery, TResult> : 
    IQueryHandler<TQuery, TResult> where TQuery : IQuery<TResult>
{
    private readonly IQueryHandler<TQuery, TResult> decorated;
    private readonly Logger logger;

    public QueryHandlerLogger(
        IQueryHandler<TQuery, TResult> decorated,
        Logger logger)
    {
        this.decorated = decorated;
        this.logger = logger;
    }

    public TResult Handle(TQuery query)
    {
        this.logger.Log(query);
        var result = this.decorated.Handle(query);
        this.logger.Log(result);
        return result;
    }
}

public class DataQueryHandlerLogger<TQuery, TResult> :
    IDataQueryHandler<TQuery, TResult> where TQuery : IQuery<TResult>
{
    private readonly IDataQueryHandler<TQuery, TResult> decorated;
    private readonly Logger logger;

    public DataQueryHandlerLogger(
        IDataQueryHandler<TQuery, TResult> decorated,
        Logger logger)
    {
        this.decorated = decorated;
        this.logger = logger;
    }

    public TResult Handle(TQuery query)
    {
        this.logger.Log(query);
        var result = this.decorated.Handle(query);
        this.logger.Log(result);
        return result;
    }
}
```

**Comment by ChristianHoe** [comment]
@qujck Hm, thats looks like a lot of repetitive code and it getting worse with each additional interface and decorator (like loggin, validation, security, ...)

@TheBigRic
> Well that's exactly what the blog about holistic abstractions is telling you! You miss an abstraction!

Maybe the problem is that I see the `IQueryHandler` not in sense of CQRS but more als an `IAOPHandler`. And so the `IMediatorHandler` just looks like an empty decorator for me.

With using inheritance in the interface like
`IMediatingHandler<TQuery, TResult> : IBusinessHandler<TQuery, TResult : IQueryHandler<TQuery, TResult>` 
and the registration like 
`container.Register(new[] { typeof(IMediatingHandler<,>), (typeof(IBusinessHandler<,>) }, new[] { typeof(IMediatingHandler<,>).Assembly }, Lifestyle.Scoped);`
`container.Register(new[] { (typeof(IBusinessHandler<,>) }, new[] { typeof(IBusinessHandler<,>).Assembly }, Lifestyle.Scoped, c => c.NotAlreadyRegistered );`
and a decorator definition like 
`LoggingDecorator<TQuery, TResult> : IBusinessHandler<,>, Common.LoggingDecorator<TQuery, TResult> : IQueryHandler<TQuery, TResult>`
 and the registration
`container.RegisterDecorator(typeof(IMediatingHandler<,>), typeof(Common.LoggingDecorator<,>), lifestyle); //using co-/contra variance on the decorator` 
there would be no need for the mediator and the separate decorator implementations.

**Comment by TheBigRic** [comment]
> Hm, thats looks like a lot of repetitive code and it getting worse with each additional interface and decorator (like loggin, validation, security, ...)

I think we have a different understanding of what `IQueryHandler` is or is not. So I give it one more try.... 

Applying cross cutting concerns is much more easy when you develop your application around some open-generic interfaces and use, what I call, a message based architecture. This gives you the ability to write one decorator per cross cutting concern per interface type. 

If you're in a place where `IQueryHandler` needs more data, because of composites or whatever, you could or should hide `IQueryHandler` behind another abstraction just as the blog about Holistic Abstraction is telling you. In that case you would typically apply the cross cutting concerns to new abstraction, `IMediationHandler` in this case. 

But even this seems not to be sufficient for you and you still want to apply e.g. a `LoggingDecorator` to the underlying `IQueryHandlers`.But why would you need that? Maybe it is just naming where all misunderstandings come from. 

Personally I use `IQueryHandler` as the root type. If I come across a use case where the same set of read operations take place I would refactor that repetative code to some service or specialized repository and the whole problem goes away.

This looks like:
``` c#
class GetData1QueryHandler : IQueryHandler<GetData1Query, Data1Result>
{
    private readonly Data1Service data1Service;
    public Data1Result Handle(GetData1Query query)
    {
        return this.data1Service.GetData(query);
    }
}

class GetData2QueryHandler : IQueryHandler<GetData2Query, Data2Result>
{
    private readonly Data2Service data2Service;
    public Data2Result Handle(GetData2Query query)
    {
        return this.data2Service.GetData(query);
    }
}

class GetCompositeQueryHandler : IQueryHandler<GetCompositeQuery, CompositeResult>
{
    private readonly Data1Service data1Service;
    private readonly Data2Service data2Service;
    public CompositeResult Handle(GetCompositeQuery query)
    {
        var data1 = this.data1Service.GetData(query.Data1QueryData);
        var data2 = this.data2Service.GetData(query.Data2QueryData);
        // Combine data1 and data2....
        return this.Combine1And2(data1, data2);
    }
}
```

Cross cutting concerns are only applied to IQueryHandler!! But if needed, and only if needed, cross *could* be applied to Data1Service.

Inheritance is almost never the solution to your design problem. In that case you'll most probably are just missing another interface. If you think the design of @qujck is not DRY enough you could imagine adding an `IPerformQueryService` with a single implementation and a single decorator for logging.

**Comment by ChristianHoe** [comment]
I got the opportunity spend some hours with refactoring an old application. The application is basically a data hub, delivering mostly plain data from different sources as a web service (via WCF). Additionally there is a small GUI.

After I got in contact with AOP in a different project and I got infested by some of the abilities (for example the traceability via logging), but I was very unhappy with the "weaving" of the aspect. And when I stumbled across steves blog entry, I though "Thats just AOP without the weaving and also typed!". I checked the old application if using `IQueryHandler` would be an option. The main structure of the application are WCF, Business, DataAccess. Changing WCF part to the new Interface would be really easy. The DataAccess part is by far read-only operations, so using this approach is also simple. And same applies to Business. So applying this new approach was looking good.

With my AOP experience back in my mind, a CommonLoggingDecorator is a must for traceability (with on/off-switch) - But, and here the trouble starts, not in the WCF part. In the WCF there must be an always-on logger like a WCFLoggingDecorator. And having so, the CommonLoggingDecorator should not be applied to avoid redundancies. So if Simple Injector could apply the decorator target-dependent everything would be fine and I am done, but it seems not to be possible.
The next best way seems to use an interface for each (sub-)part which needs a different set of decorators. I though inheritance is the way to go, because `IBusinessHandler` is just a specialized `IQueryHandler`. But when i tried to apply my CommonLogginDecorator via Simple Injector to my now specilized handlers, I got an type mismatch. My simple global traceability in mind, I started to use inheritance for the CommonLogginDecorator too.
The next escalation level happend, when I looked into Business. Trying to keep it simple, I have some public methods which are reusing some of the public methods (the composites). But doing so, applies the decorator set for the public methods twice. Thats was the Moment the `IMediatingHandler` came into the game. But getting the same class with different interfaces is not that easy in Simple Injector (see http://simpleinjector.readthedocs.org/en/latest/tornlifestyle.html ). So I had to implement a concrete kind of `IMediatingHandler` and added up another layer of abstraction and ended here because everything felt like a big workaround.

Beside the whole design discussion, it would be nice if you could think about the following feature requests:
* RegisterDecorator: context not only contains implemented type but also injected-into type (For example: it would be possible to decorate dependent in which namespace the query is injected)
* RegisterDecorator: allowing co-/contra variance on decorators so a `IQueryHandler`-decorator could be applied on a `IBusinessHandler`-query (with `IBusinessHandler : IQueryHandler`)
* Register: Allow batch registration of a type for different Interfaces (see http://simpleinjector.readthedocs.org/en/latest/tornlifestyle.html)

**Comment by dotnetjunkie** [comment]
> So if Simple Injector could apply the decorator target-dependent everything would be fine and I am done

Can you elaborate on this? What do you mean by "target-dependent"?

> RegisterDecorator: context not only contains implemented type but also injected-into type (For example: it would be possible to decorate dependent in which namespace the query is injected)

Do you mean apply the decorator conditionally based on the type the decoratee (the handler) is injected into?

> RegisterDecorator: allowing co-/contra variance on decorators so a IQueryHandler-decorator could be applied on a IBusinessHandler-query (with IBusinessHandler : IQueryHandler)
Register: Allow batch registration of a type for different Interfaces

That's not variance :). Casting an `IEnumarable<string>` to an `IEnumerable<object>` is a form of variance.

But what you want is not possible indeed. This is not a limitation in Simple Injector though. This is something you can't do by hand. Since the IQueryHandler-decorator doesn't implement IBusinessHandler, you can never inject it into a consumer that expects IBusinessHandler. Try to do this by hand, and you'll see a C# compile error.

**Comment by ChristianHoe** [comment]
> Do you mean apply the decorator conditionally based on the type the decoratee (the handler) is injected into?

Yes.

> Can you elaborate on this? What do you mean by "target-dependent"?

With:
`class A { public A(IQueryHandler handler) { ..} }`
`class B { public B(IQueryHandler handler) { ..} }`

Target-dependent registration:
`// used if decoratee is injected into A`
`container.RegisterDecorator(typeof(IQueryHandler <,>), typeof(A_Decorator<,>), lifestyle, c => c.HandlerInjectedIntoType == typeof(A));`
`// used if decoratee is injected into B`
`container.RegisterDecorator(typeof(IQueryHandler <,>), typeof(B_Decorator<,>), Lifestyle, c => c.HandlerInjectedIntoType == typeof(B));`

>That's not variance :). Casting an  `IEnumarable<string>`  to an  `IEnumerable<object>`  is a form of variance.
>But what you want is not possible indeed. This is not a limitation in Simple Injector though. This is something you can't do by hand. Since the IQueryHandler-decorator doesn't implement IBusinessHandler, you can never inject it into a consumer that expects IBusinessHandler. Try to do this by hand, and you'll see a C# compile error.

To be honest I have some trouble with variance, but what is the difference? With the following definitions:
`class A { }`
`class B { } : A`
`IValidate<in T>`

I can mix classes which implements` IValidate<A>` and `IValidate<B>` when validating a `B`. Is there really a difference in applying a lesser specified `Decorator<IQueryHandler>` on a `Decoratee<IBusinessHandler>` with `IBusinessHandler : IQueryHandler`? Just like working on a    `IEnumerable<object>` instead of an `IEnumarable<string>` ...

**Comment by dotnetjunkie** [comment]
> Is there really a difference in applying a lesser specified Decorator<IQueryHandler> on a Decoratee<IBusinessHandler> with IBusinessHandler : IQueryHandler?

Yes, there is a difference, because you didn't define a `Decorator<IQueryHandler>` in your code; you defined a `Decorator<TQuery>`. Variance applies to generic type arguments, but IQueryHandler is not used as type argument. And as I said, if you wrap an IQueryHandler interface around an IBusinessHandler, you will never be able to inject it into a consumer that expects an IBusinesHandler.
